xrtl/testing/render_test.h

```diff
@@ -0,0 +1,49 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef XRTL_TESTING_RENDER_TEST_H_
+#define XRTL_TESTING_RENDER_TEST_H_
+
+#include <string>
+
+#include "xrtl/testing/gtest.h"
+
+namespace xrtl {
+namespace testing {
+
+// Base for image pixel diffing tests.
+//
+// Usage:
+//  class MyTest : public testing::RenderTest {};
+//  TEST_F(MyTest, DoSomething) {
+//    auto pixel_data = ...;
+//    EXPECT_TRUE(CompareImages(path_to_golden, pixel_data, width, height));
+//  }
+class RenderTest : public ::testing::Test {
+ protected:
+  // Returns false if the test image and golden image are different.
+  bool CompareImages(const std::string& golden_image_path,
+                     const std::string& test_image_path);
+
+  // Returns false if the test image and golden image are different.
+  // The test image must have three components (rgb, 8 bits each) per pixel.
```

Comment:
This restriction is painful, as for example any framebuffer we read back will be 4 component rgba. We don't want to have to do format conversion in tests before calling this. Instead, having the function switch between a comparison of 1 component images against an alpha-only PNG, 3 component with RGB PNG, and 4 component with PNG with alpha would make it easier to use in a variety of scenarios.